### PR TITLE
feat(annotations) :add color options annotation

### PR DESCRIPTION
**Pull request by robertbolan-bits**
re #528 
Added the ability to select colors for annotations Headers and Body. The color of the tail should match the color selected for the body unless you are not showing the body. If you are not showing the body then the color will match the header.

### CODE REVIEW

**Comment by robertbolan-bits** [comment]
Test Steps:
Step 1: Right click on the map & select 'Create Annotation' at the bottom of the list.
Step 2: Left click on the color picker next to 'Header Background' OR 'Body Background'
Step 3: Select a color from the color palette
Step 4: Click 'OK' to save your color selections.

File: src/os/annotation/annotation.js [hunk 1]
```diff
@@ -15,7 +15,10 @@ os.annotation.DEFAULT_OPTIONS = {
   showName: true,
   showDescription: true,
   size: [200, 100],
-  offset: [0, -75]
+  offset: [0, -75],
+  headerBG: '#52575c',
+  bodyBG: '#3a3f44'
+
```
Comment (welchyd) [inline src/os/annotation/annotation.js]: Can you set the color defaults to the constants you defined later in this file? ie `os.annotation.DEFAULT_ANNOTATION_HEADER_BACKGROUND_COLOR`

File: src/os/annotation/annotationui.js [hunk 2]
```diff
@@ -53,7 +53,9 @@ os.ui.Module.directive('annotation', [os.annotation.annotationDirective]);
  */
 os.annotation.UI_TEMPLATE_ =
   '<div class="c-annotation u-hover-container">' +
-    '<svg class="c-annotation__svg"><path/></svg>' +
+    '<svg class="c-annotation__svg">' +
+      '<path ng-style="{ \'fill\': ctrl.options.showDescription ? ctrl.options.bodyBG : ctrl.options.headerBG}" />' +
```
Comment (welchyd) [inline src/os/annotation/annotationui.js]: No need to escape `fill`

File: src/os/annotation/annotation.js [hunk 3]
```diff
@@ -15,7 +15,10 @@ os.annotation.DEFAULT_OPTIONS = {
   showName: true,
```
Comment (welchyd) [inline src/os/annotation/annotation.js]: This file needs a require for the colorpicker directive

File: src/os/annotation/annotationui.js [hunk 4]
```diff
@@ -64,10 +66,12 @@ os.annotation.UI_TEMPLATE_ =
...
       '<div class="card-header flex-shrink-0 text-truncate px-1 py-0" title="{{ctrl.name}}"' +
           'ng-show="ctrl.options.showName"' +
-          'ng-class="!ctrl.options.showDescription && \'h-100 border-0\'">' +
+          'ng-class="!ctrl.options.showDescription && \'h-100 border-0\'"' +
+          'ng-style="{\'background\': ctrl.options.headerBG }">' +
         '{{ctrl.name}}' +
       '</div>' +
-      '<div class="card-body p-1 u-overflow-y-auto" ng-show="ctrl.options.showDescription">' +
+      '<div class="card-body p-1 u-overflow-y-auto" ng-show="ctrl.options.showDescription" ' +
+          'ng-style="{\'background\': ctrl.options.bodyBG }">' +
```
Comment (welchyd) [inline src/os/annotation/annotationui.js]: No need to escape `background`

**Comment by welchyd** [comment]
Live test here: http://wild-racoon.surge.sh/

File: views/annotation/annotationoptions.html [hunk 5]
```diff
@@ -35,13 +35,41 @@
...
-            ng-change="ctrl.updateAnnotation()"
+            ng-change="ctrl.updatePreview()"
             ng-disabled="!ctrl.annotationOptions.show"
             ng-model="ctrl.annotationOptions.showDescription"/>
         <label class="custom-control-label" ng-attr-for="showAnnotationDescription{{ctrl.uid}}">
           Show Description
         </label>
       </div>
     </div>
+    <div class="form-row">
```
Comment (welchyd) [inline views/annotation/annotationoptions.html]: As a space saving measure, could you move the new options to fill the extra space on the right side?

space

**Comment by robertbolan-bits** [comment]
Live Link: http://unsuitable-spy.surge.sh/

**Comment by robertbolan-bits** [comment]
my live link has been updated: http://unsuitable-spy.surge.sh/

**Comment by welchyd** [comment]
I'm digging the new changes for the annotation colors defaulting to match the theme and change with the theme.

Are the default colors supposed to show as white though? I thought the colors in the color pickers might reflect the current theme ... I guess we don't have a great way to indicate 'this is tied to the theme'
AnnotationDefaultColors

**Comment by robertbolan-bits** [comment]
They are defaulting to white because they are undefined. I havent found a good way to set those colors to the current theme.

**Comment by wallw-bits** [comment]
Can you not use the theme style classes by default and only have the custom background color override those if they are set by the user?

I can also show you a way to get the theme colors directly, but I do not think that will end up being the best solution here.

**Comment by justin-bits** [comment]
Build failing due to a timing issue with image comparison tests. PR coming soon with fix.

**Comment by justin-bits** [comment]
https://github.com/ngageoint/opensphere/pull/583

File: views/annotation/annotationoptions.html [hunk 6]
```diff
@@ -3,45 +3,72 @@
...
+            </label>
+          </div>
+        </div>
+        <div class="col">
+          <div title="Change the background color of the header">
+            <colorpicker name="headerColor"
+                ng-attr-id="headerColor{{ctrl.uid}}"
+                color="ctrl.tempHeaderBG"
+                show-reset="true"
+                title="Sets the bg color for header"
```
Comment (bradh) [inline views/annotation/annotationoptions.html]: Maybe expand bg to background?

File: views/annotation/annotationoptions.html [hunk 7]
```diff
@@ -3,45 +3,72 @@
...
+          </label>
+        </div>
+        </div>
+        <div class="col">
+          <div class="d-inline-block" title="Change the background color of the description">
+            <colorpicker name="bodyColor"
+                ng-attr-id="bodyColor{{ctrl.uid}}"
+                color="ctrl.tempBodyBG"
+                show-reset="true"
+                title="Sets the bg color for body"
```
Comment (bradh) [inline views/annotation/annotationoptions.html]: Maybe expand bg to background?

File: test/plugin/file/kml/ui/placemarkededit.test.js [hunk 8]
```diff
@@ -9,6 +9,7 @@ goog.require('plugin.places.PlacesManager');
 describe('plugin.file.kml.ui.placemarkededit', function() {
   var scope, element, rootFolder, folder1;
 
+
```
Comment (bradh) [inline test/plugin/file/kml/ui/placemarkededit.test.js]: Probably unneeded change

File: test/plugin/file/kml/ui/placemarkededit.test.js [hunk 9]
```diff
@@ -27,6 +28,7 @@ describe('plugin.file.kml.ui.placemarkededit', function() {
 
       $compile(element)(scope);
     });
+    spyOn(window, 'getComputedStyle').andReturn('<div></div>');
```
Comment (bradh) [inline test/plugin/file/kml/ui/placemarkededit.test.js]: This might be unrelated - not sure what the test change intent is here.

**Comment by wallw-bits** [comment]
@robertbolan-bits if you address the merge conflicts I think this is ready for a final test (pending the minor updates that Brad mentioned).

**Comment by schmidtk** [comment]
@robertbolan-bits The annotation options layout needed a little adjusting, but may have been due to merged changes. I submitted a suggested layout in robertbolan-bits/opensphere/pull/1.

**Pull request merged**